# GraftM

---

## Overview
Basically, GraftM is a pipeline used for identifying and classifying marker gene reads from metagenomic datasets. If you are after a tool that provide rapid and accurate analysis of short read sequence data, then you'll probably want to give GraftM a try

Head on over to the [GraftM page](http://geronimp.github.io/graftM/) for more general information.

Because GraftM is still being actively developed backwards incompatibility may happen in the future, although we will aim to minimise this as much as possible.

---

## Installation
The python code can be installed through pip:
```
pip install graftm
```
### Dependencies
GraftM requires a few extra binary applications as well, unfortunately:
* orfm v. >= 0.2.0 (https://github.com/wwood/OrfM)
* hmmer v. >= 3.1b1 (http://hmmer.janelia.org/)
* fxtract (https://github.com/ctSkennerton/fxtract)
* pplacer v. >= 2.6.32 (http://matsen.fhcrc.org/pplacer/)
* krona v. >= 2.4 (http://sourceforge.net/p/krona/home/krona/)

---

## Help + Contact
A [manual](https://github.com/geronimp/graftM/wiki) is available in the form of the wiki here on github.

If you have any further comments, complaints or recomendations about GraftM, drop an email to the [SupportM](https://groups.google.com/forum/?hl=en#!forum/supportm) public help forum.

---
Software by Joel A. Boyd (@geronimp) and [Ben J. Woodcroft](http://www.ecogenomic.org/users/ben-woodcroft) (@wwood) at the [Australian Centre for Ecogenomics](http://ecogenomic.org)
Released under GPL3 - see LICENCE.txt for licencing details
